src/lib/update-lock.ts

```diff
@@ -82,6 +82,55 @@ async function dispose(
 	}
 }
 
+/**
+ * Given a lockfile path `p`, return a tuple [appId, serviceName] of that path.
+ * Paths are assumed to end in the format /:appId/:serviceName/(resin-)updates.lock.
+ */
+function getIdentifiersFromPath(p: string) {
+	const parts = p.split('/');
+	if (parts.pop()?.match(/updates\.lock/) === null) {
+		return [];
+	}
+	const serviceName = parts.pop();
+	const appId = parts.pop();
+	return [appId, serviceName];
+}
+
+export type LocksTakenMap = Map<number, string[]>;
+
+/**
+ * Return a list of services that are locked by the Supervisor under each appId.
+ */
+export function getServicesLockedByAppId(): LocksTakenMap {
+	const locksTaken = lockfile.getLocksTaken();
```

Comment:
The function `getLocksTaken` returns only the locks taken during the current supervisor run. I know we delete existing locks during supervisor start, but I wonder if we need to remove that cleanup step, as user containers could take the lock during the supervisor startup and leave the user app in an inconsistent state. 

If that's the case, `getLocksTaken` should maybe return the actual state of locks on disk.

You may have already changed this in a future commit. Ignore this if that's the case

Reply:
> lockfile module stores locks it has taken in memory, meaning Supervisor-taken locks from an older Supervisor run wouldn't get picked up.

I understand that, maybe the update-lock module needs to check all locks in the filesystem during initialization?

> Why would the user containers taking the lock leave the user app in an inconsistent state? 

Is not necessarily likely, but here is a scenario. Supervisor has 2 services, service A is kill-then-download and service B is download-then-kill 

- Supervisor waits for locks and takes them
- Service A is killed and image is downloaded, supervisor crashes while downloading image for service B
- During restart the supervisor deletes the locks, which gives service B the chance to take the lock again
- Now the user app is stuck with one dead service and one service running until service B releases the lock again

Reply:
Thanks for explaining, that makes sense. I'll modify this in a separate commit.

Reply:
Removed SIGTERM lockfile removal, lockfile in-memory tracking, and tightened requirements for a service to be counted as locked here: https://github.com/balena-os/balena-supervisor/pull/2234/commits/99ddf8387317b23bfec1e5aaa0161cbf7992ec51